tests/console/orphaned_packages_check.pm

```diff
@@ -18,27 +18,62 @@ use warnings;
 use testapi;
 use utils 'zypper_call';
 use version_utils 'is_upgrade';
+use Data::Dumper;
+
+# Perfomring a DVD/Offline system upgrade cannot update
+# all potential packages already present on the SUT
+# A subsequent 'zypper dup' is nescessary to ensure
+# all packages available in the repo are up-to-date
+# poo#61829
+sub is_offline_upgrade {
+    return !get_var('ONLINE_MIGRATION') && is_upgrade;
+}
+
+# Compare detected orphans with whitelisted if any
+sub compare_orphans_lists {
+    my %args = @_;
+
+    # Remove heading from zypper pa output
+    shift @{$args{zypper_orphans}};
+    print "TEST " . Dumper(@{$args{zypper_orphans}}) . "\n";
+    map { s/\s+//g; $_ } @{$args{zypper_orphans}};
+    print "TEST " . Dumper(@{$args{zypper_orphans}}) . "\n";
+    record_info('Detected Orphans',    "@{ $args{zypper_orphans} }");
+    record_info('Orphans whitelisted', $args{whitelist} // 'None defined in the test suite');
+    return 0 unless ($args{whitelist});
+
+    # Remove repeating packages from the list
+    my %wl = map { $_ => 1 } (split(',', $args{whitelist}));
+
+    return 0 if ((scalar @{$args{zypper_orphans}}) > (scalar(keys %wl)));
+
+    my @not_handled_orphans = grep { !$wl{$_} } @{$args{zypper_orphans}};
+    record_info('Missing', (@not_handled_orphans) ? ("@not_handled_orphans", result => 'fail') : 'None');
+
+    return ((scalar @not_handled_orphans) == 0);
+}
 
 sub run {
     select_console 'root-console';
+
+    record_info('Upgraded?', 'Has the SUT been upgraded? Offline upgrades can possibly cause orphans',
+        result => (is_offline_upgrade) ? 'ok' : 'fail');
+
+    # Orphans are also expected on JeOS without SDK module (jeos-firstboot, jeos-license and live-langset-data)
     # Save the orphaned packages list to one log file and upload the log, so QA can use this log to report bug
-    my $cmd = 'zypper pa --orphaned | grep -v "\(release-DVD\|release-dvd\|openSUSE-release\|skelcd\)" | (! grep "@System") > /tmp/orphaned.log';
-    # there are orphans on older, unsupported openSUSE versions which we
-    # upgrade from. They will most likely never be fixed
-    my $expect_failure = is_upgrade && get_var('HDD_1') =~ /\b(1[123]|42)[\.-]/;
-    set_var('ZYPPER_ORPHANED_CHECK_ONLY', get_var('ZYPPER_ORPHANED_CHECK_ONLY', $expect_failure));
-    if (get_var('ZYPPER_ORPHANED_CHECK_ONLY')) {
-        script_run($cmd);
-    }
-    else {
-        assert_script_run($cmd, fail_message => "Orphaned packages found, set 'ZYPPER_ORPHANED_CHECK_ONLY' to only check and not fail the test");
+    my @orphans = split('\n', script_output q[zypper -qs 11 pa --orphaned | tee -a /tmp/orphaned.log | awk 'NR>2 {print $3}']);
```

Comment:
what does the `-qs 11` mean?

Reply:
```bash
✗ zypper --help                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
Global Options:                                                                                                                                                                                                                                                        
    --quiet, -q             Suppress normal output, print only error messages.                                                                                                                                                                                                
    --table-style, -s <INTEGER>                                                                                                                                               
                            Table style (0-11).                                                                                                                               
```

`-s 11` makes the table borderless, thus I do not need to handle whitespaces in each record.

Reply:
```
 ~ zypper -s 5 pa --orphaned                            
Retrieving repository 'Main Repository (DEBUG)' metadata ...............................................................................................................[done]
Building repository 'Main Repository (DEBUG)' cache ....................................................................................................................[done]
Retrieving repository 'Main Repository (Sources)' metadata .............................................................................................................[done]
Building repository 'Main Repository (Sources)' cache ..................................................................................................................[done]
Loading repository data...
Reading installed packages...
S  ┇ Repository ┇ Name                                            ┇ Version                             ┇ Arch  
┅┅┅┋┅┅┅┅┅┅┅┅┅┅┅┅┋┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┋┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┋┅┅┅┅┅┅┅
i+ ┇ @System    ┇ boost-license1_67_0                             ┇ 1.67.0-2.3                          ┇ noarch
i+ ┇ @System    ┇ edk2.git-ovmf-x64                               ┇ 0-20190704.1237.g204ae9da23         ┇ noarch
i+ ┇ @System    ┇ google-chrome-stable                            ┇ 63.0.3239.108-1                     ┇ x86_64
i+ ┇ @System    ┇ libboost_date_time1_67_0                        ┇ 1.67.0-2.3                          ┇ x86_64
i+ ┇ @System    ┇ libboost_filesystem1_67_0                       ┇ 1.67.0-2.3                          ┇ x86_64
i+ ┇ @System    ┇ libboost_iostreams1_67_0
```

Reply:
nice! *Interesting* UX for zypper ;)